### PR TITLE
Increased KCP Fragments

### PR DESCRIPTION
**Pull request by JesusLuvsYooh**
Credits to the top-men / advisors and the big guy himself.

We do get people trying to send large mb maps, or 4k HD camera footage (in Update() )  in hashtag help, so its for those rare few, + future proofing.

### CODE REVIEW

File: kcp2k/Assets/kcp2k/kcp/Kcp.cs [hunk 1]
```diff
@@ -25,8 +25,8 @@ public class Kcp
         public const int MTU_DEF = 1200;           // default MTU (reduced to 1200 to fit all cases: https://en.wikipedia.org/wiki/Maximum_transmission_unit ; steam uses 1200 too!)
         public const int ACK_FAST = 3;
         public const int INTERVAL = 100;
-        public const int OVERHEAD = 24;
-        public const int FRG_MAX = byte.MaxValue;  // kcp encodes 'frg' as byte. so we can only ever send up to 255 fragments.
+        public const int OVERHEAD = 25;
```
Comment (vis2k) [inline kcp2k/Assets/kcp2k/kcp/Kcp.cs:28]:
```suggestion
        public const int OVERHEAD = 25;         // original kcp uses 24 byte. we add 1 more byte for frg ushort max
```

File: kcp2k/Assets/kcp2k/kcp/Kcp.cs [hunk 2]
```diff
@@ -521,10 +521,10 @@ public int Input(byte[] data, int offset, int size)
                 if (conv_ != conv) return -1;
 
                 offset += Utils.Decode8u(data, offset, ref cmd);
-                // IMPORTANT kcp encodes 'frg' as 1 byte.
-                // so we can only support up to 255 fragments.
-                // (which limits max message size to around 288 KB)
-                offset += Utils.Decode8u(data, offset, ref frg);
+                // IMPORTANT kcp encodes 'frg' as 1 ushort.
+                // so we can only support up to 65355 fragments.
+                // (which limits max message size to around 70,000 KB)
```
Comment (vis2k) [inline kcp2k/Assets/kcp2k/kcp/Kcp.cs:526]:
```suggestion
                // (which limits max message size to around 70 MB)
```

File: kcp2k/Assets/kcp2k/kcp/Segment.cs [hunk 3]
```diff
@@ -30,10 +30,10 @@ internal int Encode(byte[] ptr, int offset)
             int offset_ = offset;
             offset += Utils.Encode32U(ptr, offset, conv);
             offset += Utils.Encode8u(ptr, offset, (byte)cmd);
-            // IMPORTANT kcp encodes 'frg' as 1 byte.
-            // so we can only support up to 255 fragments.
-            // (which limits max message size to around 288 KB)
-            offset += Utils.Encode8u(ptr, offset, (byte)frg);
+            // IMPORTANT kcp encodes 'frg' as 1 ushort.
+            // so we can only support up to 65355 fragments.
+            // (which limits max message size to around 70,000 KB)
```
Comment (vis2k) [inline kcp2k/Assets/kcp2k/kcp/Segment.cs:35]:
```suggestion
            // (which limits max message size to around 70 MB)
```

File: kcp2k/Assets/kcp2k/kcp/Kcp.cs [hunk 4]
```diff
@@ -25,8 +25,8 @@ public class Kcp
         public const int MTU_DEF = 1200;           // default MTU (reduced to 1200 to fit all cases: https://en.wikipedia.org/wiki/Maximum_transmission_unit ; steam uses 1200 too!)
         public const int ACK_FAST = 3;
         public const int INTERVAL = 100;
-        public const int OVERHEAD = 24;
-        public const int FRG_MAX = byte.MaxValue;  // kcp encodes 'frg' as byte. so we can only ever send up to 255 fragments.
+        public const int OVERHEAD = 25;         // original kcp uses 24 byte. we add 1 more byte for frg ushort max
```
Comment (vis2k) [inline kcp2k/Assets/kcp2k/kcp/Kcp.cs:28]:
```suggestion
        public const int OVERHEAD = 25;            // original kcp uses 24 byte. we add 1 more byte for frg ushort max
```

File: kcp2k/Assets/kcp2k/kcp/Kcp.cs [hunk 5]
```diff
@@ -25,8 +25,8 @@ public class Kcp
         public const int MTU_DEF = 1200;           // default MTU (reduced to 1200 to fit all cases: https://en.wikipedia.org/wiki/Maximum_transmission_unit ; steam uses 1200 too!)
         public const int ACK_FAST = 3;
         public const int INTERVAL = 100;
-        public const int OVERHEAD = 24;
-        public const int FRG_MAX = byte.MaxValue;  // kcp encodes 'frg' as byte. so we can only ever send up to 255 fragments.
+        public const int OVERHEAD = 25;            // original kcp uses 24 byte. we add 1 more byte for frg ushort max
```
Comment (vis2k) [inline kcp2k/Assets/kcp2k/kcp/Kcp.cs:28]:
```suggestion
        public const int OVERHEAD = 25;               // original kcp uses 24 byte. we add 1 more byte for frg ushort max
```

File: kcp2k/Assets/kcp2k/kcp/Kcp.cs [hunk 6]
```diff
@@ -25,8 +25,8 @@ public class Kcp
         public const int MTU_DEF = 1200;           // default MTU (reduced to 1200 to fit all cases: https://en.wikipedia.org/wiki/Maximum_transmission_unit ; steam uses 1200 too!)
         public const int ACK_FAST = 3;
         public const int INTERVAL = 100;
-        public const int OVERHEAD = 24;
-        public const int FRG_MAX = byte.MaxValue;  // kcp encodes 'frg' as byte. so we can only ever send up to 255 fragments.
+        public const int OVERHEAD = 25;               // original kcp uses 24 byte. we add 1 more byte for frg ushort max
```
Comment (vis2k) [inline kcp2k/Assets/kcp2k/kcp/Kcp.cs:28]:
```suggestion
        public const int OVERHEAD = 25;              // original kcp uses 24 byte. we add 1 more byte for frg ushort max
```

**Comment by vis2k** [comment]
looks good, just need one 70 MB send test to be sure.
needs different window sizes in test settings tho, I can add it later

**Comment by vis2k** [comment]
for future reference.
70mb sending did not seem to work atm.
worth considering keeping the old limit possibly

**Comment by JesusLuvsYooh** [comment]
Old limit was 288 kb right?
I can get 10,000 kb to work with the above changes, before unity, kcp or my hardware disagrees with it.
Its a massive difference, even if we cant reach the 70,000 kb we'd hoped.

I'l ask in the Discord for opinions.

**Comment by vis2k** [comment]
not merged yet, sorry.
considering NetworkReader/Writer fixed capacity pooling.
where we would always pool max sized readers/writers.
in which case the 144 KB max is quite nice.
int.max nice would not be so ni ce

**Pull request closed without merging**